overlord/snapstate/snapmgr.go

```diff
@@ -404,28 +399,45 @@ func (m *SnapManager) ensureRefreshes() error {
 
 	tr := config.NewTransaction(m.state)
 
-	// allow disabling auto-refresh in the tests
-	if osutil.GetenvBool("SNAPD_DEBUG") {
-		var refreshDisabled bool
-		err := tr.Get("core", "refresh.disabled", &refreshDisabled)
-		if err != nil && !config.IsNoOption(err) {
-			return err
-		}
-		if refreshDisabled {
-			return nil
-		}
+	// get last refresh time
+	var lastRefresh time.Time
+	err := m.state.Get("last-refresh", &lastRefresh)
+	if err != nil && err != state.ErrNoState {
+		return err
 	}
 
-	var lastRefresh time.Time
-	err := tr.Get("core", "refresh.last", &lastRefresh)
+	// get our schedule
+	refreshScheduleStr := defaultRefreshSchedule
+	err = tr.Get("core", "refresh.schedule", &refreshScheduleStr)
 	if err != nil && !config.IsNoOption(err) {
 		return err
 	}
+	refreshSchedule, err := timeutil.ParseSchedule(refreshScheduleStr)
+	if err == nil {
+		err = refreshScheduleUsesWeekdays(refreshSchedule)
+	}
+	if err != nil {
+		logger.Noticef("cannot use refresh.schedule: %s", err)
+		refreshSchedule, err = timeutil.ParseSchedule(defaultRefreshSchedule)
+		if err != nil {
+			panic(fmt.Sprintf("defaultRefreshSchedule cannot be parsed: %s", err))
+		}
+		tr.Set("core", "refresh.schedule", defaultRefreshSchedule)
+		tr.Commit()
+	}
 
-	nextRefresh := lastRefresh.Add(minRefreshInterval).Add(m.refreshRandomness)
-	if time.Now().Before(nextRefresh) {
-		return nil
+	// already have a refresh timer
+	if m.nextRefresh != nil {
+		if m.currentRefreshSchedule == refreshScheduleStr {
+			return nil
+		}
+		// the refresh schedule has changed
+		logger.Debugf("Option refresh-schedule changed, reloading.")
+		m.nextRefresh.Stop()
+		m.nextRefresh = nil
+		m.lastRefreshAttempt = time.Time{}
```

Comment:
also not sure:

	// Check that we have reasonable delays between unsuccessful attempts.
	// If the store is under stress we need to make sure we do not
	// hammer it too often
	if !m.lastRefreshAttempt.IsZero() && m.lastRefreshAttempt.Add(10*time.Minute).After(time.Now()) {
		return nil
	}

makes sense where it is now, time.Now() is not when we will run the refresh attempt, it's just the time we are trying to schedule it, should it be moved inside the AfterFunc callback? or the logic tweaked some other way?